Firebase/Database/Utilities/FUtilities.m

```diff
@@ -122,54 +122,81 @@ + (NSString *) decodePath:(NSString *)pathString {
+    if (queryParamIndex != NSNotFound) {
+        path = [path substringToIndex:queryParamIndex];
     }
 
-    host = [[url substringToIndex:slashIndex] lowercaseString];
-    if (slashIndex >= url.length) {
-        url = @"";
-    } else {
-        url = [url substringFromIndex:slashIndex + 1];
+    return path;
+}
+
++ (FParsedUrl *) parseUrl:(NSString *)url {
+    // For backwards compatibility, support URLs without schemes on iOS.
+    if (![url containsString:@"://"]) {
+        url = [@"http://" stringByAppendingString:url];
+    }
+
+    NSString* originalPathString = [self extractPathFromUrlString:url];
+
+    // Sanitize the database URL by removing the path component, which may contain invalid URL characters.
+    NSString* sanitizedUrlWithoutPath = [url stringByReplacingOccurrencesOfString:originalPathString withString:@""];
+    NSURLComponents *urlComponents = [NSURLComponents componentsWithString:sanitizedUrlWithoutPath];
+    if (!urlComponents) {
+        [NSException raise:@"Failed to parse database URL" format:@"Failed to parse database URL: %@", url];
     }
 
-    NSArray *parts = [host componentsSeparatedByString:@"."];
+    NSString* host = [urlComponents.host lowercaseString];
+    NSString* namespace;
+    bool secure;
+
+    if (urlComponents.port != nil) {
+        secure = [urlComponents.scheme isEqualToString:@"https"];
+        host = [host stringByAppendingFormat:@":%@", urlComponents.port];
+    } else if ( [urlComponents.host isEqualToString:@"localhost"]) {
+        secure = NO;
+    } else {
+        secure = YES;
+    };
+
+    NSArray *parts = [urlComponents.host componentsSeparatedByString:@"."];
     if([parts count] == 3) {
-        NSInteger colonIndex = [[parts objectAtIndex:2] rangeOfString:@":"].location;
-        if (colonIndex != NSNotFound) {
-            // we have a port, use the provided scheme
-            secure = [scheme isEqualToString:@"https"];
-        } else {
-            secure = YES;
+        namespace = [parts[0] lowercaseString];
+    } else {
+        // Attempt to extract namespace from "ns" query param.
+        NSArray *queryItems = urlComponents.queryItems;
+        for (NSURLQueryItem *item in queryItems) {
+            if ([item.name isEqualToString:@"ns"]) {
+                namespace = item.value;
+                break;
+            }
         }
 
-        namespace = [[parts objectAtIndex:0] lowercaseString];
-        NSString* pathString = [self decodePath:[NSString stringWithFormat:@"/%@", url]];
-        path = [[FPath alloc] initWith:pathString];
-    }
-    else {
-        [NSException raise:@"No Firebase database specified." format:@"No Firebase database found for input: %@", url];
+        if (!namespace) {
+            namespace = [parts[0] lowercaseString];
+        }
     }
 
+
+    NSString* pathString = [self decodePath:[NSString stringWithFormat:@"/%@", originalPathString]];
+    FPath *  path = [[FPath alloc] initWith:pathString];
```

Comment:
seems like there are extra spaces here?  Why does clang-format not fix that?

Reply:
Fixed manually. Seems like we are excluded from the formatter (but it still takes 10 seconds to run). I'll get the formatting enabled for RTDB in a follow up.

Reply:
This source is now clang-formatted.